File: src/main/java/me/andre111/d20server/command/WhisperCommand.java
package me.andre111.d20server.command;

import me.andre111.d20common.model.chat.ChatEntry;
import me.andre111.d20common.model.profile.Profile;
import me.andre111.d20server.service.ChatService;
import me.andre111.d20server.service.UserService;

public class WhisperCommand extends Command {

	public WhisperCommand(String name, String[] aliases) {
		super(name, aliases);
	}

	@Override
	public void execute(Profile profile, String arguments) {
		String[] split = arguments.split(" ", 2);
		if(split.length != 2) {
			ChatService.appendError(profile, "Usage: /whisper <name> <message>");
			return;
		}
		
		String name = split[0].toLowerCase();
		String message = split[1];
		
		// find receiver
		Profile reciever = null;
		for(Profile other : UserService.getAllProfiles()) {
			if(name.equals(other.getName().toLowerCase())) {
				reciever = other;
			}
		}
		if(reciever == null) {
			ChatService.appendError(profile, "Unknown player: "+split[0]);
			return;
		}
		
		// build message
		StringBuilder sb = new StringBuilder();
		sb.append("<p class=\"chat-sender chat-sender-special\">");
		sb.append(ChatService.escape(profile.getName()));
		sb.append(" to ");
		sb.append(ChatService.escape(reciever.getName()));
		sb.append(": ");
		sb.append("</p>");
		sb.append("<p class=\"chat-message\">");
		sb.append(ChatService.escape(message));
		sb.append("</p>");
		
		// determine recipents
		long[] recipents = new long[] { profile.id(), reciever.id() };
		
		// append message
		ChatService.append(true, new ChatEntry(sb.toString(), profile.id(), false, recipents));
	}

}

File: core/server/command/whisper-command.js
import { ChatEntry } from '../../common/message/chat/chat-entry.js';
import { I18N } from '../../common/util/i18n.js';
import { splitArguments } from '../../common/util/stringutil.js';
import { ChatService } from '../service/chat-service.js';
import { UserService } from '../service/user-service.js';
import { Command } from './command.js';

export class WhisperCommand extends Command {
    constructor(name, aliases) {
        super(name, aliases, false);
    }

    execute(profile, args) {
        const split = splitArguments(args, 2);
        if (split.length < 2) throw new Error(I18N.get('commands.error.arguments', 'Wrong argument count: %0', '<name> <message>'));

        const name = split[0].toLowerCase();
        const message = split[1];

        // find receiver
        const reciever = UserService.findByUsername(name, true);
        if (!reciever) throw new Error(`Unknown player: ${split[0]}`);

        // parse message
        const parsed = ChatService.parseInlineExpressions(message, profile);

        // build message
        var text = '<div class="chat-sender chat-sender-special">' + I18N.get('command.whisper.title', '%0 to %1: ', ChatService.escape(profile.getUsername()), ChatService.escape(reciever.getUsername())) + '</div>';
        text = text + '<div class="chat-message">' + parsed.string + '</div>';

        // determine recipents
        const recipents = [profile.getID(), reciever.getID()];

        // append message
        ChatService.append(true, new ChatEntry(text, profile.getID(), false, recipents, parsed.diceRolls, parsed.triggeredContent));
    }
}
